tools/code_analyzer/op_dependency.cpp

```diff
@@ -0,0 +1,396 @@
+      if (gvInit->isNullValue()) continue;
+      ConstantDataArray* gvInitArray = dyn_cast<ConstantDataArray>(gvInit);
+      if (!gvInitArray || !gvInitArray->isCString()) continue;
+      std::string gvInitStr = gvInitArray->getAsCString().str();
+      if (!isFunctionSchemaString(gvInitStr)) continue;
+      std::string schemaStr = truncateFunctionSchemaString(gvInitStr);
+
+      // Track all seen schema strings. Useful for checking whether we miss any
+      // dependency in the logic below.
+      keyNodes->insert(schemaStr);
+
+      // Search for GV users.
+      for (auto gvUser : GV.users()) {
+        // Search for "i8* getelementptr inbounds" ConstantExpr.
+        ConstantExpr* CE = dyn_cast<ConstantExpr>(gvUser);
+        if (!CE) continue;
+
+        // Search for CE users.
+        for (auto ceUser : CE->users()) {
+          // Search for "std::basic_string" constructor.
+          Instruction* ceUserInst = dyn_cast<Instruction>(ceUser);
+          if (!ceUserInst) continue;
+          analyzeFunctionSchemaUser(
+              schemaStr, *ceUserInst, schemaStrToFunctions,
+              functionsToSchemaStrs);
+        }
+      }
+    }
+  }
+
+  // Use fuzzy match to find all llvm::Function instances related to the
+  // function schema string. There are two types of relationship:
+  // 1) function schema to function (op registration);
+  // 2) function to function schema (op invocation);
+  //
+  // For op invocation, simply check whether the parent function calls
+  // 'c10::Dispatcher::findSchema' since the use of the schema string.
+  //
+  // For op registration, collect all function references between two following
+  // "c10::RegisterOperators::Options::schema" calls. The function references
+  // can be direct function call, function pointer or template argument.
+  //
+  // Note that these rules highly depend on how LLVM emit IRs. If it changes
+  // the sequence of schema string reference and c10::RegisterOperators API call
+  // then we might need use more sophisticated alias analysis, analyze clang-AST
+  // instead or simplify leverage the connection between function schema string
+  // and related function name (as they are generated by script).
+  static void analyzeFunctionSchemaUser(
+      const std::string& schemaStr,
+      Instruction& I,
+      GRAPH* schemaStrToFunctions,
+      GRAPH* functionsToSchemaStrs) {
+    std::string parentFunctionName = I.getFunction()->getName();
+    bool seenOptionsSchemaCall = false;
+    for (Instruction* cur = getNextInstruction(I);
+         cur;
+         cur = getNextInstruction(*cur)) {
+      // Check pattern to call registered ops.
+      if (checkInstructionCalledFunction(*cur, "c10::Dispatcher::findSchema")) {
+        (*functionsToSchemaStrs)[parentFunctionName].insert(schemaStr);
+        continue;
+      }
+
+      // Check pattern to register ops.
+      if (checkInstructionCalledFunction(
+          *cur, "c10::RegisterOperators::Options::schema")) {
```

Comment:
The name of this function matters, there should be a comment on its definition saying that if you rename/renamespace it, you MUST update the static analysis too.

Reply:
I guess it probably is a good idea to comment that `schema` must point to exact string; you're not allowed to indirect it through some other mechanism.

Reply:
I think it's fine to have some indirection as long as it's inlined as I don't assume schema() is directly following the string reference. For example: "schema(inline_function("aten::acos"))" might be unfolded into:
```
@.str.100 = "aten::acos"
@.str.101 = "aten::asin"
%5 = allocate basic_string_1
%6 = allocate basic_string_2
// registering aten::acos:
%10 = invoke std::basic_string ... %5, @.str.100 # construct basic_string_1
%11 = inline_function_instructions %5 # use basic_string_1
%12 = invoke schema(... %11)
...
%20 = store ptrtoint @at::acos # associate with function ptr
...
// registering aten::asin:
%30 = invoke std::basic_string ... %6, %.str.101 # construct basic_string_2
%31 = .... %30 # use basic_string_2
%32 = invoke schema(... %31)
...
%40 = store ptrtoint @at::asin
...
```
For "aten::acos", it will collect all function pointers between %12 and %32.
It doesn't matter whether it's the same global string constant "aten::acos" across all units as I use the actual string as the key of my hash map.
But if %5, %10 are in a different function which is not inlined with schema() call, then this method will fail.

Reply:
Alternatively I could also scan all "Options::schema" function usages and check all its operands to search for all string constant references which look like function schemas. But it might be a little bit overkill at this point. We can cover these corner cases later.

Reply:
Yes, I agree if the function gets inlined away this code will work. It's OK to not handle the corner cases, I just think it's good to explicitly say what all of the corner cases are.